### DISCUSSION TITLE
GPIO output doesn't show up in /sys/class/gpio  and GPIO inputs are destroyed at program exit

### DISCUSSION THREAD

**Discussion by gitterman** [gpiozero/gpiozero#1067] [Q&A]
Operating system: Raspbian Buster
Python version:  3.7
Pi model: Pi Model B r2
GPIO Zero version: 1.5.1
Pin factory used: RPiGPIO

If I define an input button like   'btn = Button( 3 )'  I get its control element in /sys/class/gpio/gpio3, and I can also check the input from outside my python script (e.g. shell, ssh)
Alas, for an LED output I do not see any items in /sys/class/gpio and thus cannot control or check the output from outside my python program.

**_How does gpiozero control the LED, and how can I control it from outside of my python program?_**

Moreover, if I create GPIO controls from outside my python script like:
```
echo 4 > /sys/class/gpio/export ; echo  in > /sys/class/gpio/gpio4/direction ; echo 1  > /sys/class/gpio/gpio4/active_low
echo 3 > /sys/class/gpio/export ; exho out > /sys/class/gpio/gpio3/direction ; echo 1  > /sys/class/gpio/gpio3/value
```
then, the input structure gets destroyed at termination of the python script, whereas the output definition survives.
**_How can I prevent the automatic cleanup in gpiozero?_**

**Comment by bennuttall** [comment]
> How can I prevent the automatic cleanup in gpiozero?

Unfortunately that's not possible in gpiozero at present, see #707

@waveform80 anything to add on the other points?

**Reply by lurch** [reply]
> how can I control it from outside of my python program?

That's not the "usage model" that gpiozero uses - in the "gpiozero world" it expects to have exclusive control of the GPIOs it's using, and doesn't expect other processes to be changing the GPIOs "behind its back".

If you **really** need to be able to control GPIOs from multiple different processes at the same time, you'll probably need to modify your python program to also read and write `/sys/class/gpio/...` directly, rather than using gpiozero.

**Reply by gitterman** [reply]
> That's not the "usage model" that gpiozero uses - in the "gpiozero world" it expects to have exclusive control of the GPIOs it's using, and doesn't expect other processes to be changing the GPIOs "behind its back".

Thanks for the hint and the clarification.
What irritates me, is that inputs which are defined with gpiozero show up in /sys/class/gpio whereas outputs do not.
Also, input definitions are removed from /sys/class/gpio when the python program terminates whereas **outputs created with the sysfs interface till live there, but they are changed to inputs**!   This is a quite odd behaviour.

**Reply by lurch** [reply]
> What irritates me, is that inputs which are defined with gpiozero show up in /sys/class/gpio whereas outputs do not.

I _think_ that might depend on the particular [pin factory](https://gpiozero.readthedocs.io/en/stable/api_pins.html) that you're using? I.e. some pin factories might not create _any_ files in `/sys/class/gpio/...`, even for input pins.

**Reply by waveform80** [reply]
@lurch is correct -- this is dependent on the pin factory. In your case, you're using the default pin factory, which is RPi.GPIO and this is simply "how RPi.GPIO works". If you want the full and gory details, I'd suggest reading [a post I wrote](https://waldorf.waveform.org.uk/2021/the-pins-they-are-a-changin.html) a while back but the tl;dr version is: RPi.GPIO bangs on the hardware registers directly for as much as it possibly can (ignoring sysfs) but for edge detection on inputs it can't get away with that and has to use sysfs to avoid butting heads with the Linux kernel (which also wants to use those same registers).

Incidentally, that post also goes through how GPIO access is (very slowly) changing, and that in future this will no longer work (tl;dr the sysfs GPIO interface has been deprecated for a very long time and is scheduled for removal ... erm ... "sometime after 2020"). The new method of GPIO access (`ioctl`s against `/dev/gpiochipN` devices) automatically "cleans up" when GPIO controlling processes exit. However, "cleans up" in this instance just means that ownership handles are cleaned up by the kernel; the GPIO modes and states don't change implicitly when processes using `/dev/gpiochipN` devices exit.

However, gpiozero will continue to "clean up" (i.e. reset modes and states) on exit in the new regime too. This choice was quite deliberate, and goes back to it being intended in the first instance for education. I attended quite a few educational jams before working on gpiozero, and on several occasions witnessed young students building robots with RPi.GPIO, only to have their code crash and, because the GPIO outputs were left "on" when their code crashed, having their robots charge headlong off the table before plunging to their doom on the floor below (in one case this destroyed one of the sensors on the robot, in another it managed to destroy the Pi itself)! This is the primary reason behind trying to reset everything before exit in gpiozero -- it's typically less destructive when your code crashes (especially when motors are involved!).

That said, I do acknowledge that for some (typically more advanced) use-cases it's undesirable. At some point, there should probably be some property on the pin factory instance that can be toggled to disable the clean-up behaviour. However, I'm satisfied that the current behaviour should remain the default.

**Reply by gitterman** [reply]
Thanks a lot for the profound explanation.
I agree, that things should always default to the safe side, but I am no friend of "automagic" without an option to disable it.
Too much built it automatism make people lazy to think about security, cleanup and robustness in their programs.
Especially with robots, this can become dangerous.
So it might be a good learning experience that robots can cause damage if a program crashes ;-)

**Reply by lurch** [reply]
This conversation reminds me of the RC car that I had when I was a teenager... it had a separate (big) battery-pack for driving the motor and a smaller battery pack for powering the receiver and servos. There was a couple of times where I was unlucky enough to have the servo-battery die before the motor-battery died, and the car was driving forwards at the time, so I had to go running after it because I couldn't move the servo back to the neutral position :rofl:
